### PR TITLE
add break statement

### PR DESCRIPTION
**Pull request by qadan**
**JIRA Ticket**: https://jira.duraspace.org/browse/ISLANDORA-2009

# What does this Pull Request do?
Adds a `break` statement to the 'set' context of `NewFedoraDatastream::checksumMagicProperty()` so that setting the datastream checksum is not immediately followed up by unsetting it.

# What's new?
Applying values to `NewFedoraDatastream::checksum` should result in those values being accessible at `NewFedoraDatastream::checksum`.

Applying a correct checksum to `NewFedoraDatastream::checksum` and calling `FedoraObject::ingestDatastream()` on that datastream should result in no change.

Applying an incorrect checksum to `NewFedoraDatastream::checksum` and calling `FedoraObject::ingestDatastream()` on that datastream should result in a `RepositoryException` with a 500 code, as opposed to ignoring the given checksum and allowing one to be assigned by Fedora.

Similarly, applying an incorrect checksum to `NewFedoraDatastream::checksum` and calling `NewFedoraObject::ingestDatastream()` on that datastream and then calling `FedoraRepository::ingestObject()` on the object should also result in a `RepositoryException` with a 500 code.

# How should this be tested?
As this is pretty low-level, unit tests are being provided that check the above three cases. A tester is welcome to check this in the wild though:

```php
// From an Islandora context with devel installed:
$conn = islandora_get_tuque_connection();
$obj = $conn->repository->constructObject('test');
$ds = $obj->constructDatastream('TEST', 'M');
$ds->content = 'foo';
$ds->checksumType = 'MD5';
// This is obviously not a valid MD5 of 'foo'.
$ds->checksum = 'bar';
try {
  $obj->ingestDatastream($ds);
  $conn->repository->ingestObject($obj);
  dpm("We shouldn't be able to get here; ingesting a datastream with a mismatch is an exceptional case");
}
catch (RepositoryException $e) {
  dpm("That checksum ain't right");
}
```

Updating with this PR should give you the message in the caught `RepositoryException` section; before then, you get an ingested object that completely ignores the given checksum.

# Additional Notes:
This is an old, old bug (5+ years). The tuque documentation will have to be updated to say that checksums can be mutable. I cannot, for the life of me, envision a use case that relies on `NewFedoraDatastream::checksum` being immutable, but ... who knows.

Code that is currently applying checksums to `NewFedoraDatastream`s and ingesting them into `FedoraObjects` now throws `RepositoryException`s. Any code that was previously doing this simply failed silently (yikes ... checksums being provided by client code but not acknowledged by tuque?); it will now have to `catch` and handle the exception.

# Interested parties
This is pretty entrenched in there, so @Islandora/7-x-1-x-committers should be aware, as well as the maintainers @DiegoPino and @willtp87

### CODE REVIEW

**Comment by mjordan** [comment]
I will test this.

**Comment by DiegoPino** [comment]
I restarted the failing travis-ci test. @qadan what will happen if checksum is not passed at all, not set or not needed (like when using "R" or "E" datastram types). see https://github.com/Islandora/tuque/pull/158/files#diff-9ce53062851398d31726743f74fe5f2cR850 Will this fail when sending the info to Fedora?  So, do we need to calculate the checksum everytime? Or we have some conditional that avoids all the fuzz if NULL (don´t see that in the pull, maybe it was already there, in that case you can discard this).

**Comment by qadan** [comment]
Yep @DiegoPino, looks like in any case we shouldn't be passing along the 'none' checksum. Ignoring if it hasn't been set.

File: Object.php [hunk 1]
```diff
@@ -851,6 +851,9 @@ public function ingestDatastream(&$ds) {
       // Assume NewFedoraDatastreams will have a log message set.
       'logMessage' => ($ds instanceof NewFedoraDatastream) ? $ds->logMessage : "Copied datastream from {$ds->parent->id}.",
     );
+    if ($ds->checksum !== 'none') {
```
Comment (adam-vessey) [inline Object.php]: I think an [`isset()`](https://github.com/Islandora/tuque/pull/158/files#diff-d9ededc5334628e2c15e2f1934472aa9R678) check would suffice here, without having to worry about the value when "unset".

**Comment by mjordan** [comment]
Did some testing, and I'm not getting what I expected. I'm using the attached script, which is an expanded version of @qadan 's snippet above. To run it, put it in your tuque library directory. Then, run `php pr-158-test.php [good|bad]` (where 'good' is to use the correct checksum, 'bag' is to use a bogus one) within the 1.x branch and the qadan:1.x-ISLANDORA-2009 branch.

TL;DR is that with a valid checksum, qadan:1.x-ISLANDORA-2009 is not ingesting the object.

Here's the output of all four runs of my script. These first two runs, in branch 1.x, confirm the bug:

```
vagrant@islandora:/var/www/drupal/sites/all/libraries/tuque$ git checkout 1.x
Switched to branch '1.x'.
vagrant@islandora:/var/www/drupal/sites/all/libraries/tuque$php pr-158-test.php good
You are in branch 1.x. Regardless of checksum, object will be created and we'll see this message.
Object changeme:27 created!
vagrant@islandora:/var/www/drupal/sites/all/libraries/tuque$ php pr-158-test.php bad
You are in branch 1.x. Regardless of checksum, object will be created and we'll see this message.
Object changeme:28 created!
```
The next two runs, in branch qadan:1.x-ISLANDORA-2009, show that if the checksum is bad, the exception is caught, but that if the checksum is good, the same thing happens (and the object is not ingested):

```
vagrant@islandora:/var/www/drupal/sites/all/libraries/tuque$ git checkout 1.x-ISLANDORA-2009 
Switched to branch '1.x-ISLANDORA-2009'
Your branch is up-to-date with 'qadan/1.x-ISLANDORA-2009'.
vagrant@islandora:/var/www/drupal/sites/all/libraries/tuque$ php pr-158-test.php good
You are in branch 1.x-ISLANDORA-2009, and the checksum is good. The object was not ingested, so the PR need further testing.
vagrant@islandora:/var/www/drupal/sites/all/libraries/tuque$ php pr-158-test.php bad
You are in branch 1.x-ISLANDORA-2009, if checksum is bad. The object was not ingested, which is what we want.
```

Can someone else confirm that when the checksum is valid, qadan/1.x-ISLANDORA-2009' does not ingest the object? It could very well be that my script's logic is flawed, so unless you have another way of testing this PR, I'd confirm that my script is not wrong.

[pr-158-test.php.txt](https://github.com/Islandora/tuque/files/1180639/pr-158-test.php.txt)

**Comment by adam-vessey** [comment]
@mjordan: I'm not sure your MD5 is correct:
```bash 
$ echo foo | md5
d3b07384d113edec49eaa6238ad5ff00
$ echo -n foo | md5
acbd18db4cc2f85cedef654fccc4a4d8
```

The first includes a newline after the string `foo`, so it's calculating the hash for the string `foo\n`.
The `-n` instructs `echo` to avoid outputting the newline after the string being echo'd.

**Comment by DiegoPino** [comment]
@mjordan have not tested with the latest changes, but a quick look at your test code tells me it should be ok, maybe the MD5 is wrong? What about running the actual MD5 php function?
. Side note: any one has seen this ?
Wonder if this https://groups.google.com/forum/#!topic/fedora-tech/2IvZWxqJ2gI can be related, or can probably make other stuff fail on out side?

**Comment by mjordan** [comment]
Gak! Google let me down! I generated that hash using `md5sum <<<"foo"`.

@adam-vessey and @DiegoPino I just reran my script with the correct checksum and all is :+1: . Boy that is embarrassing.

**Comment by DiegoPino** [comment]
@mjordan all good, no embarrassing at all. Thanks a lot for testing.  You are ++

**Comment by willtp87** [comment]
I think this is looking good.

**Comment by mjordan** [comment]
I'm happy to merge tomorrow first thing, to give 24 hours.

**Review by mjordan — APPROVED** [review]
Smoke tests pass.

**Comment by qadan** [comment]
Awesome, thanks!

**Comment by mjordan** [comment]
Travis is failing on a few builds, it's phpunit. Here is the phpunit output on the failures:

```
Tests: 181, Assertions: 978, Failures: 1, Skipped: 2, Incomplete: 8.
Tests: 183, Assertions: 982, Failures: 1, Skipped: 2, Incomplete: 8.
Tests: 183, Assertions: 976, Failures: 1, Skipped: 2, Incomplete: 8, Risky: 5.
Tests: 183, Assertions: 980, Failures: 1, Skipped: 2, Incomplete: 8, Risky: 5.
```

**Comment by willtp87** [comment]
Should be green now.

**Comment by mjordan** [comment]
Thanks. What caused the failures?

**Comment by willtp87** [comment]
I think the logs go away once the tests are re-run, but I think there may be racy things in the tests.

**Comment by mjordan** [comment]
@rosiel @DiegoPino Travis prevented my merge on July 18, any objections to merging now?

**Comment by DiegoPino** [comment]
@mjordan I see it merged.. where do you see that failure in doing so?
https://github.com/Islandora/tuque/commit/6fd377372a52286b8d203c1e81b16677fb6504e5

**Comment by mjordan** [comment]
Was looking at https://github.com/Islandora/tuque/pull/158#issuecomment-318737637 - didn't notice it was merged. Do you want to close the JIRA ticket or want me to?

**Comment by DiegoPino** [comment]
and closed as fixed. Probably github was a bit crazy that day? A failing Travis-CI after a merge does not avoid the merge. Had no fix version (bad.. bad..) but that got corrected in the JIRA also. Will appear now in the release notes.
